<a href="https://colab.research.google.com/github/songqsh/ma2071_v01/blob/master/src/asymptotic_discrete_dynamic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Asymptotic of discrete dynamics

## Abstract

Consider dynamic given by
$$x_{k+1} = A x_k$$
- introduce concepts of atrractor, repeller, saddle point
- how to classify these behavior
- Sec 5.6

## Analysis

$x_k$ is usually understood as the quantity of certian object at time $k$. 

__ex__

For instance, $x_k$ is the population of owl and rat at $k$-th months in the predator-prey system of Section 5.6.  Given that 
the generator 
$$A = \begin{bmatrix}
.5 & .4 \\
-.104 & 1.1
\end{bmatrix}
$$
Find its long-term behavior.

__strategy__

If $x_0 = \sum_i c_i v_i$, where $v_i$ is eigenvector of $\lambda_i$, then we have
$$x_k = A^k x_0 = \sum_i c_i \lambda_i^k v_i.$$
As $k\to \infty$, 
- if $|\lambda_i|>1$, then the $i$-th term goes to $\infty$
- if $|\lambda_i|<1$, then the $i$-th term goes to $0$.

In [0]:
from sympy import init_printing, Matrix, symbols, eye, printing
def custom_latex_printer(exp,**options):
    from google.colab.output._publish import javascript
    url = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.3/latest.js?config=default"
    javascript(url=url)
    return printing.latex(exp,**options)
init_printing(use_latex="mathjax",latex_printer=custom_latex_printer)
lamda = symbols('lamda') # Note that lambda is a reserved word in python, so we use lamda (without the b)

In [5]:
A = Matrix([[.5, .4], [-0.104, 1.1]]); A

⎡ 0.5    0.4⎤
⎢           ⎥
⎣-0.104  1.1⎦

In [6]:
A.eigenvects()

⎡⎛         ⎡⎡5.0⎤⎤⎞  ⎛         ⎡⎡0.769230769230769⎤⎤⎞⎤
⎢⎜0.58, 1, ⎢⎢   ⎥⎥⎟, ⎜1.02, 1, ⎢⎢                 ⎥⎥⎟⎥
⎣⎝         ⎣⎣1.0⎦⎦⎠  ⎝         ⎣⎣       1.0       ⎦⎦⎠⎦

Note that $\lambda_1 = 1.02$ and $\lambda_2 = .58$. So, if $x_0 = c_1 v_1 + c_2 v_2$, then 
$$x_k = c_1 \lambda_1^k v_1 + c_2 \lambda_2^k v_2.$$
- If $c_1 \neq 0$, then $x_k \to \infty$
- If $c_1 = 0$, then $x_k \to 0$.

__Def__

- $0$ is called as __attractor__ if $x_k \to 0$ regardless of the initial position;
- $0$ is called as __repller__ if $x_k \to \infty$ regardless of the initial position;
- $0$ is called as __saddle point__ if $o$ attracts some solutions but repells other solutions.

In the above example, $O$ is a saddle point since
- $0$ attracts the solution starting from $x_0$ linearly dependent to $v_2$.
- $0$ repells the solution starting from any intial state linearly independent to $v_2$.